packages/dd-trace/src/appsec/api_security_sampler.js

```diff
@@ -1,61 +1,79 @@
 'use strict'
 
+const TTLCache = require('@isaacs/ttlcache')
+const web = require('../plugins/util/web')
 const log = require('../log')
+const { AUTO_REJECT, USER_REJECT } = require('../../../../ext/priority')
 
-let enabled
-let requestSampling
+const MAX_SIZE = 4096
 
-const sampledRequests = new WeakSet()
+let enabled
+let sampledRequests
 
 function configure ({ apiSecurity }) {
   enabled = apiSecurity.enabled
-  setRequestSampling(apiSecurity.requestSampling)
+  sampledRequests = new TTLCache({ max: MAX_SIZE, ttl: apiSecurity.sampleDelay * 1000 })
 }
 
 function disable () {
   enabled = false
+  sampledRequests?.clear()
 }
 
-function setRequestSampling (sampling) {
-  requestSampling = parseRequestSampling(sampling)
-}
+function sampleRequest (req, res, force = false) {
+  if (!enabled) return false
+  if (isSampled(req, res)) return false
 
-function parseRequestSampling (requestSampling) {
-  let parsed = parseFloat(requestSampling)
+  const rootSpan = web.root(req)
+  if (!rootSpan) return false
 
-  if (isNaN(parsed)) {
-    log.warn(`Incorrect API Security request sampling value: ${requestSampling}`)
+  let priority = getSpanPriority(rootSpan)
+  if (!priority) {
+    rootSpan._prioritySampler?.sample(rootSpan)
+    priority = getSpanPriority(rootSpan)
+  }
+
+  if (priority === AUTO_REJECT || priority === USER_REJECT) {
+    return false
+  }
 
-    parsed = 0
-  } else {
-    parsed = Math.min(1, Math.max(0, parsed))
+  if (force) {
+    const key = computeKey(req, res)
+    if (!key || sampledRequests.has(key)) return false
+    sampledRequests.set(key)
+    return true
   }
 
-  return parsed
+  return true
 }
 
-function sampleRequest (req) {
-  if (!enabled || !requestSampling) {
-    return false
-  }
+function isSampled (req, res) {
+  const key = computeKey(req, res)
+  if (!key) return false
```

Comment:
when there has been a problem computing the key, should we return true to avoid sampling the request?
```suggestion
  if (!key) return true
```

Reply:
@iunanua Do you think this is still relevant after the last commit?

Reply:
nop, resolving :)